src/cipher.rs

```diff
@@ -0,0 +1,89 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use dusk_jubjub::JubJubAffine;
+use rand_core::{CryptoRng, RngCore};
+
+use aes_gcm::{
+    aead::{Aead, AeadCore, KeyInit},
+    Aes256Gcm, Key,
+};
+
+const NONCE_SIZE: usize = 12;
+
+/// Size of the extra encryption data required by the
+/// cipher: the nonce (12 bytes) and the tag (16 bytes)
+pub const EXTRA_DATA_SIZE: usize = NONCE_SIZE + 16;
+
+/// Encrypts a plaintext given a shared DH secret key, returning a vector
+/// containing a nonce and the ciphertext (which includes the tag)
+pub fn encrypt<R: RngCore + CryptoRng, const ENCRYPTION_SIZE: usize>(
+    secret_key: &JubJubAffine,
+    plaintext: &[u8],
+    rng: &mut R,
+) -> [u8; ENCRYPTION_SIZE] {
+    // To encrypt using AES256 we need 32-bytes keys. Thus, we use
+    // the 32-bytes serialization of the 64-bytes DH key.
+    let key = secret_key.to_bytes();
+    let key = Key::<Aes256Gcm>::from_slice(&key);
+
+    let cipher = Aes256Gcm::new(key);
+    let nonce = Aes256Gcm::generate_nonce(rng);
+    let mut ciphertext = cipher
+        .encrypt(&nonce, plaintext.as_ref())
+        .expect("Encryption failed.");
+
+    let mut encryption = nonce.to_vec();
+    encryption.append(&mut ciphertext);
+
+    encryption.try_into().unwrap()
+}
+
+/// Decrypts an encryption (nonce + ciphertext) given a shared DH secret key,
+/// returning the plaintext
+pub fn decrypt<const PLAINTEXT_SIZE: usize>(
+    secret_key: &JubJubAffine,
+    encryption: &[u8],
+) -> [u8; PLAINTEXT_SIZE] {
+    // To decrypt using AES256 we need 32-bytes keys. Thus, we use
+    // the 32-bytes serialization of the 64-bytes DH key.
+    let key = secret_key.to_bytes();
+    let key = Key::<Aes256Gcm>::from_slice(&key);
+
+    let nonce = &encryption[..NONCE_SIZE];
+    let ciphertext = &encryption[NONCE_SIZE..];
+
+    let cipher = Aes256Gcm::new(key);
+    cipher
+        .decrypt(nonce.into(), ciphertext.as_ref())
+        .expect("Decryption failed.")
```

Comment:
What are the reasons that the decryption can fail? If they can happen with normal use, it might be a better idea to return a `Result<[u8; PLAINTEXT_SIZE], Error> here so that the user can decide on how to recover from the error instead of the entire program simply panicing.